SEED-platform/seed
# Add 2 factor authentication to login

#### Any background context you want to provide?

#### What's this PR do?
Uses an open source, well rated, free library [django-two-factor-auth](https://github.com/jazzband/django-two-factor-auth/) to add token authentication via any authenticator app (google authenticator, microsoft authenticator, etc...) or via email.

Users will only be prompted to enable 2FA on their first login following the introduction of this code. Opting in or out is at the user level and users can always access the 2FA settings from /profile  > Two Factor.

If a user chooses "Token Generator" then a QR code is displayed and can be scanned by any number of auth apps. Once scanned, users can enter the token to login

If a user chooses "Email" then a token will be sent via email to the associated account.

The UI for this library uses Django templates and does not enter the angular layer. To customize the Django templates they have been imported into the seed/landing/templates/two_factor directory and edited. To make navigation changes to the response from django-2-f-a, the response must be caught in a custom django view (see /CustomLoginView.py)

#### How should this be manually tested?
1. With token generator
Login with any user > Enable 2Fa > Select "Token Generator "
Scan QR code with auth app of choice and enter token 
User should be logged in 

2. With email
Go to profile > Two Factor > disable two factor authentication 
then re-enable two factor and select "Email". An email will be sent out, if testing locally it will likely be found in the seed_web logs.
Enter the token
User should be logged in 

<img width="732" alt="Screenshot 2024-05-03 at 9 47 25 AM" src="https://github.com/SEED-platform/seed/assets/58446472/9938ea46-0f46-4f98-b704-c2a6f839003d">

3. Disable 2fa 
Go to profile > two factor > disable two factor authentication 
log out 
log in 
user should be logged in without 2FA prompts.
 

#### What are the relevant tickets?
#4657

#### Screenshots (if appropriate)
